src/cls/IPM/General/History.cls

```diff
@@ -0,0 +1,479 @@
+{
+    $$$ThrowOnError(..%Reload())
+    Set ..Package = pName
+    Quit ..%Save()
+}
+
+Method Finalize(status As %Status, dev As %Boolean = 0) As %Status
+{
+    $$$ThrowOnError(..%Reload())
+    Set ..Success = status
+    Set ..Committed = (dev = 1) || $$$ISOK(status)
+    Set ..TimeEnd = $ZDateTime($Now(), 3)
+    If ..Package = "" {
+        Return $$$ERROR($$$GeneralError, "Package name not set")
+    }
+    Return ..%Save()
+}
+
+/// @API.Query
+/// Get load/install/uninstall records in the specified namespace
+Query Records(ascend As %Boolean, limit As %Integer, namespace As %String) As %Query(CONTAINID = 1, ROWSPEC = "Action:%String,Package:%String,Version_Major:%Integer,Version_Minor:%Integer,Version_Patch:%Integer,Version_Prerelease:%String,Version_Build:%String,SourceName:%String,SourceMoniker:%String,SourceDetails:%String,TimeStart:%String,TimeEnd:%String,NameSpace:%String,UserName:%String,Success:%String,Committed:%String,CommandString:%String") [ SqlProc ]
+{
+}
+
+ClassMethod RecordsExecute(ByRef qHandle As %Binary, ascend As %Boolean, limit As %Integer, namespace As %String) As %Status [ Internal ]
+{
+    Try {
+        Kill filter
+        Set qHandle = ..GetHistory(.filter, .ascend, .limit, .namespace)
+    } Catch e {
+        Return e.AsStatus()
+    }
+    Return $$$OK
+}
+
+ClassMethod RecordsFetch(ByRef qHandle As %Binary, ByRef Row As %List, ByRef AtEnd As %Integer = 0) As %Status [ Internal, PlaceAfter = RecordsExecute ]
+{
+    Try {
+        If qHandle.%Next() {
+            Set Row = $ListBuild(
+                qHandle.%Get("Action"),
+                qHandle.%Get("Package"),
+                qHandle.%Get("Version_Major"),
+                qHandle.%Get("Version_Minor"),
+                qHandle.%Get("Version_Patch"),
+                qHandle.%Get("Version_Prerelease"),
+                qHandle.%Get("Version_Build"),
+                qHandle.%Get("SourceName"),
+                qHandle.%Get("SourceMoniker"),
+                qHandle.%Get("SourceDetails"),
+                qHandle.%Get("TimeStart"),
+                qHandle.%Get("TimeEnd"),
+                qHandle.%Get("NameSpace"),
+                qHandle.%Get("UserName"),
+                qHandle.%Get("Success"),
+                qHandle.%Get("Committed"),
+                qHandle.%Get("CommandString")
+            )
+            Set AtEnd = 0
+        } Else {
+            Set Row = ""
+            Set AtEnd = 1
+        }
+    } Catch ex {
+        Return ex.AsStatus()
+    }
+    Return $$$OK
+}
+
+ClassMethod RecordsClose(ByRef qHandle As %Binary) As %Status [ Internal, PlaceAfter = RecordsFetch ]
+{
+    Set qHandle = ""
+    Quit $$$OK
+}
+
+/// Get the history of all installations, uninstalls, and loads in given namespace
+/// The filter argument is a multidimensional array with structure 
+/// filter(columnName) = value
+/// Where value can optionally start with >, >=, <, <=, =, <> or contain * to indicate a wildcard
```

Comment:
Rather than value starting with such things, perhaps make value a $ListBuild? The possible issue with it starting with that is that the property's value may actually be using one of these symbols.

Reply:
That's a good idea, my concern is that It can be cumbersome for users to construct a command string that contains `$ListBuild()`s. They would need to do something like
```objectscript
zpm "history find -DTimeStart=" _ $lb(">", "2000-01-01 00:00:00")
```
And I can't think of a way for it to work in interactive mode:
```
USER>zpm

=============================================================================
|| Welcome to the Package Manager Shell (ZPM). Version:                    ||
|| Enter q/quit to exit the shell. Enter ?/help to view available commands ||
|| Current registry https://pm.community.intersystems.com                  ||
=============================================================================
zpm:USER>history find -DTimeStart=???
```